src/box/tx_memory.c

```diff
@@ -0,0 +1,195 @@
+#include <assert.h>
+#include <lib/core/say.h>
+#include "tx_memory.h"
+#include "small/mempool.h"
+#include "small/region.h"
+#include "histogram.h"
+#include "txn.h"
+
+const char *TXN_ALLOC_TYPE_STRS[] = {
+	"STATEMENTS",
+	"SAVEPOINTS",
+	"USER DATA",
+	"REDO LOGS",
+	"TRIGGERS",
+	"TIMERS",
+	"JOURNAL ENTRIES"
+};
+
+static_assert(lengthof(TXN_ALLOC_TYPE_STRS) == TXN_ALLOC_MAX,
+	      "TXN_ALLOC_TYPE_STRS does not match TXN_ALLOC_TYPE");
+
+static void
+tx_track_allocation(struct tx_memory_manager *stat, struct txn *txn,
+		    uint64_t alloc_size, size_t alloc_type, bool deallocate)
```

Comment:
Shouldn't alloc_type be some kind of enum?

Reply:
The problem is `tx_memory_manager` is a base class for all tx memory managers, and each derived memory manager will use different alloc types.

Reply:
Also I'd split it into two functions: track_allocation and track_deallocation so that split these two cases and avoid relying on deallocate extra arg.